### PR TITLE
WinPB: install Cygwin using a cache

### PR DESCRIPTION
**Pull request by Willsparker**
Fixes: #1210 
Ref: #1231 , #1219 

The cached used to install Cygwin in the #1219 was flawed. The cache was created using 7-Zip which apparently doesn't retain symlinks when creating archives. In addition, extracting the archive using 7-Zip can cause issues with the paths to the symlinks , i.e. `/etc/ssl/certs/` symlinked to `\etc\pki\tls\certs\` would go to `/etc/ssl/certs` to `C:/etc/pki/tls/certs` for some reason.
To get round this, @sxa had the idea to use `tar` and `xz` built by cygwin ( located at https://ci.adoptopenjdk.net/userContent/winansible/cygwin64/bin/ ), which can be extracted by 7-Zip, which can then be used to extract the main archive.
I've been able to prove the playbook works with https://ci.adoptopenjdk.net/view/Tooling/job/VagrantPlaybookCheck/801/ , however the build failed, however seemingly not due to a cygwin issue. Therefore, I'll keep this PR in draft until I can get an example of a working build.

### CODE REVIEW

**Comment by Willsparker** [comment]
Right, finally managed to get a JDK built with a cached cygwin install, however it was locally. Ready to Review

**Comment by gdams** [comment]
I've been using this cache file in https://github.com/AdoptOpenJDK/openjdk-infrastructure/pull/1513 and the tar command is insanely slow... I'm wondering if we can make it more efficient or opt for a tool that can extract in parallel?

**Comment by aahlenst** [comment]
pigz does that well. It's parallel gzip. If size is not an issue, ISOs should retain symlinks.

**Comment by Willsparker** [comment]
Converting to draft to look into using `pigz` instead, and also replace the use of 7-Zip with:
```
Expand-Archive -LiteralPath C:\temp\cyg_toolset.zip -DestinationPath C:\temp;
```
As suggested by @gdams

**Comment by Willsparker** [comment]
https://ci.adoptopenjdk.net/view/Tooling/job/VagrantPlaybookCheck/838/ 
Ready for review - The addition of `pigz` and using `Expand Archive` worked and reduced the cygwin install down to about 18 minutes, as opposed to the 30-35 minutes it takes for a standard install.

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml [hunk 1]
```diff
@@ -15,88 +15,37 @@
...
   tags: cygwin
 
-- name: Download Cygwin installer
+- name: Retrieve cygwin cache from ci.adoptopenjdk.net
   win_get_url:
-    url: https://cygwin.com/setup-x86_64.exe
-    dest: 'C:\temp\cygwin.exe'
-    force: no
-    checksum: 4dd4d4531e8e63ade849daaaf587ba1c1430368701772c8ee42a27f4e8c373e4
+    url: https://ci.adoptopenjdk.net/userContent/winansible/cygwin64.tar.gz
```
Comment (aahlenst) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:20]: Is it obvious where the tarball comes from and how it's created?

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml [hunk 2]
```diff
@@ -15,88 +15,37 @@
...
-# grep version.
-###############################################################################
-- name: Download grep-2.27-2
+- name: Retrieve cygwin toolset from ci.adoptopenjdk.net
   win_get_url:
-    url: "{{ Cygwin_SITE_URL + '/x86_64/release/grep/grep-2.27-2.tar.xz' }}"
-    dest: "{{ Cygwin_ROOT_INST_DIR + '\\tmp\\grep.tar.xz' }}"
-    force: no
-    checksum: 36ae2483f68b3c5a4bd93814fa4f7e2576dbe6559d4ffa0589c2159e85de91f6
+    url: https://ci.adoptopenjdk.net/userContent/winansible/cygwin64/bin/*zip*/bin.zip
```
Comment (aahlenst) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:29]: `*zip*` looks suspicious. Is there some wildcard matching involved?

**Review by aahlenst — COMMENTED** [review]
My first playbook review. Please ignore the review if not helpful.

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml (hunk 2 above)
Comment (Willsparker) [reply ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:29]: That was just the link given to me when I clicked 'all files in zip' at https://ci.adoptopenjdk.net/userContent/winansible/cygwin64/ . I would assume it's just the way Jenkins generates the link, I don't think it's any wildcard matching :-)

Comment (gdams) [reply ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:29]: Yeah that `*zip*` puts all the binaries located in https://ci.adoptopenjdk.net/userContent/winansible/cygwin64/bin/ into a single zip file.

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml (hunk 1 above)
Comment (Willsparker) [reply ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:20]: I've been trying to keep a step by step in the linked issue (#1210) , which includes creating the cache- however, I'll write a comment explaining how it was made and link it above the task to make it more obvious

**Review by aahlenst — APPROVED** [review]
For me it's good to go.

**Comment by sxa** [comment]
Interesting - i wouldn't have thought pigz would provide such a significant boost in this scenario compared to the download and disk writing that's done as part of this

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml (hunk 2 above)
Comment (sxa) [reply ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:30]: I'd prefer us to just create a zip file of the files required to bootstrap this and use that. When I started on this i didn't expect there to be so many files that were going to be required :-)

**Review by sxa — COMMENTED** [review]
I can't help feel we could save a fair bit of additional time by streaming this i.e. not saving the intermediate file locally although when I tried this on Windows it kept choking part way through the download ...

**Comment by Willsparker** [comment]
@sxa I'm more than happy to try - I'll create a git branch off this PR to do some testing, and post links to `VPC.sh` runs 👍 
How were you 'streaming' the cache? I can start there

**Comment by sxa** [comment]
`curl https://ci.adopt.../cygwin64.tar.gz | gzip -d | tar xf -` (or replace `curl` with `wget -O` if preferred) ideally from a bash shell of some sort but I think I tried in various shells with no success :-(

**Comment by Willsparker** [comment]
Ah okay- the following `.dll`s are missing that stop me from starting the following `.exes`
```
bash - cygreadline7.dll
curl - cygcurl-4.dll
wget - cyggnutls-30.dll
```

**Review by sxa — APPROVED** [review]
On the basis this has been shown to be quicker I'll approve it, but I hope we can get it to stream the download/extract more efficiently at some point ...

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml [hunk 3]
```diff
@@ -15,88 +15,38 @@
...
```
Comment (sxa) [inline ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml:30]:
```suggestion
    url: https://ci.adoptopenjdk.net/userContent/winansible/cygwin64_bootstrap.zip
```

**Review by sxa — CHANGES_REQUESTED** [review]
Correcetion: If this change goes in to use the zip then I'll approve :-)

File: ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml (hunk 3 above)
Comment (Willsparker) [reply ansible/playbooks/AdoptOpenJDK_Windows_Playbook/roles/cygwin/tasks/main.yml]: URL & Checksum updated with most recent commit :+1:

**Review by sxa — APPROVED** [review]
Approving, but let's do a final validate through VPC before you merge this :-)

**Comment by Willsparker** [comment]
https://ci.adoptopenjdk.net/job/VagrantPlaybookCheck/861/ on it :-) This will fail due to the current MSVS_2019 issue ( #1526 ), but that's after the cygwin install, so is still a viable check.

**Comment by Willsparker** [comment]
The above `VPC` run failed - The issue is that the `cygwin_bootstrap` zip is not unpacking into a folder, but I'll confirm this on a VM (Once I've got my laptop setup for VMs ... )

**Comment by Willsparker** [comment]
Rerunning now : https://ci.adoptopenjdk.net/job/VagrantPlaybookCheck/862/

**Comment by Willsparker** [comment]
https://ci.adoptopenjdk.net/job/VagrantPlaybookCheck/862/
This went through okay, so I think it's good to merge :-)

**Pull request merged**